### PR TITLE
Use a shim to invoke initialBuildSteps for ghci #1364

### PR DESCRIPTION
**Pull request by mgsloan**
Here's how this works:

* Outputs a `StackSetupShim.hs` file to the build's temporary directory.

* Builds both simple setup and custom setup files using this file as an input, along with `-main-is: StackSetupShim.mainOverride`.

* When we want to run the `initialBuildSteps` cabal function, we run the setup exe with the arguments `repl stack-initial-build-steps`.  This condition is special-cased by the shim, and causes the function we want to be run.  Any other invocations of the executable get passed through to the standard simple / default setup.

In summary, this is a bit of a hack, but I think it works perfectly in all cases we care about.

### CODE REVIEW

**Comment by mgsloan** [comment]
Pinging @chrisdone - this should be a major improvement for intero as well!  What do you think of this approach?

**Comment by chrisdone** [comment]
Sounds good. I looked through the patch. I'm not too fresh on Stack's codebase at the moment for it to be a good review, but the approach is okay (with the one problem that anyone's custom build repl hooks won't be launched, it seems to me) and yet will yield a smoother user experience.

**Comment by mgsloan** [comment]
Cool!  I think this approach is solid, so merging.

> with the one problem that anyone's custom build repl hooks won't be launched, it seems to me

Yeah, if somehow the user uses the repl hook to affect the initialBuildSteps, then that won't be applied. I don't anyone does that, though.  Ordinary repl commands / hook will still work, just not integrated with `stack ghci`.
